### ISSUE TITLE
CrossentropyCategorical1Hot() not on GPU

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#5036]
I have see a case where CrossentropyCategorical1Hot was not on the GPU. We don't have it implemented, just CrossentropyCategorical1Hot with the softmax was implemented on the GPU.

It happen that there is something between the softmax and the crossentropy. This currently make this not working on the GPU.

As we implemented logsoftmax on the CPU and GPU, I think we can just implement CrossentropyCategorical1Hot on the GPU and we will support those cases while staying numericaly stable?

**Comment by lamblin** [comment]
That Op is not really needed, as it does not provide numerical stability. In fact, it could actually prevent LogSoftmax from being inserted.
It is only useful as an intermediate step before the bigger CrossEntropySoftmax... gets inserted, and maybe for the special case of advanced indexing it is performing.
I think we should replace it by `-log(coding_dist[arange(true_one_of_n.shape[0]), true_one_of_n])` or something like that.
We would need to wait for advanced indexing on the GPU for that, though

**Comment by lamblin** [comment]
It is actually still useful when `coding_dist` is not the output of a softmax.
I have a PR in preparation for that.
